rpcs3/Emu/Cell/Modules/cellAtracXdec.cpp

```diff
@@ -0,0 +1,869 @@
+				{
+					notify_error.cbFunc(ppu, CELL_ADEC_ERROR_ATX_FATAL, notify_error.cbArg);
+				}
+			}
+
+			// LLE does not initialize the output address if parsing the ATS header fails
+			vm::ptr<void> output = vm::null;
+			u32 decoded_samples_num = 0;
+
+			if (error != CELL_ADEC_ERROR_ATX_ATSHDR)
+			{
+				// The LLE SPU thread would crash if you attempt to decode without a valid configuration
+				ensure(decoder.config_is_set, "Attempted to decode with invalid configuration");
+
+				output.set(work_mem.addr() + atracXdecGetSpursMemSize(decoder.nch_in));
+
+				const auto au_start_addr = atracx_param.au_includes_ats_hdr_flg == CELL_ADEC_ATRACX_ATS_HDR_INC ? cmd.au_start_addr.get_ptr() + sizeof(AtracXdecAtsHeader) : cmd.au_start_addr.get_ptr();
+				std::memcpy(work_mem.get_ptr(), au_start_addr, decoder.nbytes);
+
+				if (int err = avcodec_send_packet(decoder.ctx, decoder.packet); err)
+				{
+					cellAtracXdec.error("avcodec_send_packet() failed (err=0x%x='%s')", err, utils::av_error_to_string(err));
+					error = CELL_ADEC_ERROR_ATX_NON_FATAL; // Not accurate, FFmpeg doesn't provide detailed errors like LLE
+				}
+
+				// Always call avcodec_receive_frame() to clear the AVFrame, even if avcodec_send_packet() returned an error
+				if (int err = avcodec_receive_frame(decoder.ctx, decoder.frame); err)
+				{
+					cellAtracXdec.error("avcodec_receive_frame() failed (err=0x%x='%s')", err, utils::av_error_to_string(err));
+					error = CELL_ADEC_ERROR_ATX_NON_FATAL; // Not accurate, FFmpeg doesn't provide detailed errors like LLE
+				}
+
+				decoded_samples_num = decoder.frame->nb_samples;
+				ensure(decoded_samples_num == 0u || decoded_samples_num == ATXDEC_SAMPLES_PER_FRAME);
+
+				// The first frame after a starting a new sequence or after an error is replaced with silence
+				if (skip_next_frame && error == CELL_OK)
+				{
+					skip_next_frame = false;
+					decoded_samples_num = 0;
+					std::memset(output.get_ptr(), 0, ATXDEC_SAMPLES_PER_FRAME * (decoder.bw_pcm & 0x7full) * decoder.nch_out);
+				}
+
+				// Convert FFmpeg output to LLE output
+				const auto output_f32 = vm::static_ptr_cast<f32>(output);
+				const auto output_s16 = vm::static_ptr_cast<s16>(output);
+				const auto output_s32 = vm::static_ptr_cast<s32>(output);
+
+				switch (decoder.bw_pcm)
+				{
+				case CELL_ADEC_ATRACX_WORD_SZ_FLOAT:
+					for (u32 channel_idx = 0; channel_idx < decoder.nch_in; channel_idx++)
+					{
+						for (u32 sample_idx = 0; sample_idx < decoded_samples_num; sample_idx++)
+						{
+							const f32 sample = reinterpret_cast<f32*>(decoder.frame->data[channel_idx])[sample_idx];
+
+							if (sample >= std::bit_cast<f32>(std::bit_cast<u32>(1.f) - 1))
+							{
+								output_f32[sample_idx * decoder.nch_in + ATXDEC_AVCODEC_CH_MAP[decoder.ch_config_idx - 1][channel_idx]] = std::bit_cast<be_t<f32>>("\x3f\x7f\xff\xff"_u32); // Prevents an unnecessary endian swap
```

Comment:
You can move `ATXDEC_AVCODEC_CH_MAP[decoder.ch_config_idx - 1][channel_idx]` out of all the inner loops